src/psyclone/tests/psyir/backend/visitor_test.py

```diff
@@ -138,12 +139,66 @@ def test_psyirvisitor_visit_arg_error():
 
     '''
     visitor = PSyIRVisitor(indent_string=" ", initial_indent_depth=4)
-    with pytest.raises(VisitorError) as excinfo:
+    with pytest.raises(TypeError) as excinfo:
         visitor("hello")
+    assert ("The PSyIR visitor functor method only accepts a PSyIR Node "
+            "as argument, but found 'str'." in str(excinfo.value))
+    with pytest.raises(VisitorError) as excinfo:
+        visitor._visit("hello")
     assert ("Visitor Error: Expected argument to be of type 'Node' but found "
             "'str'." in str(excinfo.value))
 
 
+def test_psyirvisitor_lower_dsl_concepts():
+    ''' Test that DSL concepts are lowered by the visitors but the node is not
+    modified after the visitor has finished. '''
+
+    class MyDSLNode(Statement):
+        ''' DSL Concept that lowers to a return statement '''
+        _text_name = "MyDSLNode"
+
+        def lower_to_language_level(self):
+            ''' MyDSLNode lowers to a return statement and adds a symbol
+            if it is inside an scoping region. '''
+            self.scope.symbol_table.add(DataSymbol("val", REAL_TYPE))
+            self.replace_with(Return())
+
+    class MyVisitor(PSyIRVisitor):
+        ''' Simple Visitor for Schedules and Return statements '''
+        def return_node(self, _):
+            ''' Return node visitor '''
+            return "return"
+
+        def schedule_node(self, node):
+            ''' Schedule node visitor '''
+            return "schedule(" + self._visit(node.children[0]) + ")"
+
+    # Create a custom visitor and dsl-level node
+    visitor = MyVisitor(indent_string=" ", initial_indent_depth=4)
+    schedule = Schedule()
+    my_dsl_node = MyDSLNode()
+    schedule.addchild(my_dsl_node)
+
+    # Visit DSL Node in a tree (the tree should not be modified)
+    assert visitor(schedule) == "schedule(return)"
+    assert isinstance(schedule.children[0], MyDSLNode)
+    assert schedule.children[0] is my_dsl_node
+    assert len(schedule.symbol_table.symbols) == 0
+
+    # Visit DSL Node directly (the tree is also not modified)
+    assert visitor(my_dsl_node) == "return"
+
```

Comment:
Please add some checks here that the tree is not modified after the Visit.

Reply:
Done, I was meaning to do that, I even wrote it in the comment :)